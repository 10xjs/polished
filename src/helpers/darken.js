// @flow

import parseColorString from '../internalHelpers/_parseColorString'
import rgbToHsl from '../internalHelpers/_rgbToHsl'
import hslToHex from '../internalHelpers/_hslToHex'
import hslToRgb from '../internalHelpers/_hslToRgb'

/**
 * Darken
 */
function darken(amount: number, color: string): string {
  // parse color string to hsl
  const hslColor = rgbToHsl(parseColorString(color))
  const value = {
    ...hslColor,
    lightness: Math.max(0, Math.min(1, hslColor.lightness - amount)),
  }
  return value.alpha >= 1 || value.alpha === undefined
    ? hslToHex(value.hue, value.saturation, value.lightness)
    : `rgba(${hslToRgb(value.hue, value.saturation, value.lightness)},${value.alpha})`
}

export default darken
