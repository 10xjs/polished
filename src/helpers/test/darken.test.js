// @flow
import darken from '../darken'

describe('rgb', () => {
  it('should darken a color by 20%', () => {
    expect(darken(0.2, '#444')).toMatchSnapshot()
  })

  it('should darken a typical color with opacity by 10%', () => {
    expect(darken(0.1, 'rgba(220,140,140,0.7)')).toMatchSnapshot()
  })

  it('should darken a color with a value of 255 and opacity by 10%', () => {
    expect(darken(0.1, 'rgba(255,140,140,0.7)')).toMatchSnapshot()
  })
})
