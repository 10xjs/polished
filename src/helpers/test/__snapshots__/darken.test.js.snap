exports[`rgb should darken a color by 20% 1`] = `"#111"`;

exports[`rgb should darken a color with a value of 255 and opacity by 10% 1`] = `"rgba(255,89,89,0.7)"`;

exports[`rgb should darken a typical color with opacity by 10% 1`] = `"rgba(208,101,101,0.7)"`;
